### ISSUE TITLE
Crash when loading screenlogicpy

### ISSUE THREAD

**Issue by American1337** [dieselrabbit/screenlogicpy#30]
Hi, I had good luck with the pentair integration for HA until my pool contractor made a few changes to the system (he changed the names of the lights, water features, etc).

Any idea what's going on with this log?

```
2023-01-22 15:00:57.315 ERROR (MainThread) [homeassistant] Error doing job: Fatal error: protocol.data_received() call failed.
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/requests/utility.py", line 41, in takeMessages
    message, pos = getSome(f"{msgLength}s", data, pos)
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/requests/utility.py", line 69, in getSome
    return struct.unpack_from(fmt, buff, offset)[0], newoffset
struct.error: unpack_from requires a buffer of at least 1076 bytes for unpacking 1068 bytes at offset 8 (actual buffer size is 1024)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/selector_events.py", line 868, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/requests/protocol.py", line 45, in data_received
    for messageID, messageCode, message in takeMessages(data):
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/requests/utility.py", line 45, in takeMessages
    raise ScreenLogicError(
screenlogicpy.const.ScreenLogicError: Unexpected amount of data received. Data: b'\x03\x00\xf50,\x04\x00\x00d\x00\x00\x00(h(h\x00\x02\x00\xa00\x00\x00\x00\x0e\x00\x00\x00Water Features\x00\x00\x1b\x00\x00\x00\xf4\x01\x00\x00\x03\x00\x00\x00Spa\x00G\x01\x01\x01\x00\x00\x00\x01\xd0\x02\x00\x00\xf5\x01\x00\x00\x07\x00\x00\x00Cleaner\x00\x15\x05\x00\x00\x00\x00\x00\x02\xd0\x02\x00\x00\xf6\x01\x00\x00\x04\x00\x00\x00Jets-\x00\x01\x00\x00\x00\x00\x03\xd0\x02\x00\x00\xf7\x01\x00\x00\r\x00\x00\x00Water Feature\x00\x00\x00S\x00\x02\x00\x00\x00\x00\x04\xd0\x02\x00\x00\xf8\x01\x00\x00\x0b\x00\x00\x00Fiber Optic\x00\x1d\x0b\x04\x00\x00\x00\x00\x05\xd0\x02\x00\x00\xf9\x01\x00\x00\x04\x00\x00\x00Pool<\x02\x00\x01\x00\x00\x00\x06\xd0\x02\x00\x00\xfa\x01\x00\x00\x0b\x00\x00\x00Color Wheel\x00\x16\x0c\x03\x00\x00\x00\n\x07\xd0\x02\x00\x00\xfb\x01\x00\x00\x0b\x00\x00\x00Color Wheel\x00\x16\x0c\x03\x00\x00\x00\n\x08\xd0\x02\x00\x00\xfc\x01\x00\x00\x05\x00\x00\x00Aux 7\x00\x00\x00\x08\x00\x05\x00\x00\x00\x00\t\xd0\x02\x00\x00\xfd\x01\x00\x00\x05\x00\x00\x00Aux 8\x00\x00\x00\t\x00\x05\x00\x00\x00\x00\n\xd0\x02\x00\x00\xfe\x01\x00\x00\t\x00\x00\x00Spa Light\x00\x00\x00I\n\x03\x00\x00\x00\x00\x0b\xd0\x02\x00\x00\xff\x01\x00\x00\x0b\x00\x00\x00Pool Light \x00\\\t\x03\x00\x00\x00\x00\x0c\xd0\x02\x00\x00\x00\x02\x00\x00\x0b\x00\x00\x00Pool Light \x00]\t\x03\x00\x00\x00\x00\r\xd0\x02\x00\x00\x01\x02\x00\x00\x0b\x00\x00\x00Pool Light \x00^\t\x03\x00\x00\x00\x00\x0e\xd0\x02\x00\x00\x02\x02\x00\x00\x0b\x00\x00\x00Pool Light \x00_\t\x03\x00\x00\x00\x00\x0f\xd0\x02\x00\x00\x03\x02\x00\x00\x0b\x00\x00\x00Pool Light \x00`\t\x03\x00\x00\x00\x00\x10\xd0\x02\x00\x00\x04\x02\x00\x00\t\x00\x00\x00Waterfall\x00\x00\x00U\x00\x02\x00\x00\x00\x00\x11\xd0\x02\x00\x00\x05\x02\x00\x00\x05\x00\x00\x00Aux 8\x00\x00\x00\t\x00\x05\x00\x00\x00\x00\x12\xd0\x02\x00\x00\x06\x02\x00\x00\x05\x00\x00\x00Aux 9\x00\x00\x00\n\x00\x05\x00\x00\x00\x00\x13\xd0\x02\x00\x00\x07\x02\x00\x00\x06\x00\x00\x00Aux 10\x00\x00\x0b\x00\x05\x00\x00\x00\x00\x14\xd0\x02\x00\x00\x08\x02\x00\x00\x08\x00\x00\x00Upr Poola\x00\x02\x00\x00\x00\x00\x15\xd0\x02\x00\x00\t\x02\x00\x00\x0b\x00\x00\x00Upr Cleaner\x00b\x00\x00\x00\x00\x00\x00\x16\xd0\x02\x00\x00\n\x02\x00\x00\x0b\x00\x00\x00Upr Wtrfall\x00c\x00\x02\x00\x00\x00\x00\x17\xd0\x02\x00\x00\x0b\x02\x00\x00\x0b\x00\x00\x00U Pool Ligh\x00d\x07\x04\x00\x00\x00\x00\x18\xd0\x02\x00\x00\x0c\x02\x00\x00\x05\x00\x00\x00Aux 5\x00\x00\x00\x06\x00\x05\x00\x00\x00\x00\x19\xd0\x02\x00\x00\x1c\x02\x00\x00\x05\x00\x00\x00Slide\x00\x00\x00E\x00\x02\x00\x00\x00\x00)\xd0\x02\x00\x00\x1d\x02\x00\x00\r\x00\x00\x00Spa Waterfall\x00\x00\x00M\x0e\x02\x00\x00\x00\x00*\xd0\x02\x00\x00\x08\x00\x00\x00\x05\x00\x00\x00White\x00\x00\x00\xff\x00\x00\x00\xff\x00\x00\x00\xff\x00\x00\x00\x0b\x00\x00\x00Light Green\x00\xa0\x00\x00\x00\xff\x00\x00\x00\xa0\x00\x00\x00\x05\x00\x00\x00Green\x00\x00\x00\x00\x00\x00\x00\xff\x00\x00\x00P\x00\x00\x00\x04\x00\x00\x00Cyan\x00\x00\x00\x00\xff\x00\x00\x00\xc8\x00\x00\x00\x04\x00\x00\x00Blued\x00\x00\x00\x8c\x00\x00\x00\xff\x00\x00\x00\x08\x00\x00\x00Lavender\xe6\x00\x00\x00\x82\x00\x00\x00\xff\x00\x00\x00\x07\x00\x00\x00Magenta\x00\xff\x00\x00\x00\x00\x00\x00\x00'
2023-01-22 15:01:02.280 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Pentair: 20-17-74 for screenlogic
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/requests/config.py", line 12, in async_request_pool_config
    await asyncio.wait_for(
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 372, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/__init__.py", line 60, in async_setup_entry
    await gateway.async_connect()
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/gateway.py", line 77, in async_connect
    await self._async_get_config()
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/gateway.py", line 176, in _async_get_config
    await async_request_pool_config(self.__protocol, self.__data)
  File "/usr/local/lib/python3.10/site-packages/screenlogicpy/requests/config.py", line 24, in async_request_pool_config
    raise ScreenLogicWarning("Timeout polling pool config")
screenlogicpy.const.ScreenLogicWarning: Timeout polling pool config

```

**Comment by dieselrabbit** [comment]
Thanks for the detailed logs.

Could you confirm your expected circuit names?

**Comment by American1337** [comment]
Here is screenshot of the names in the app

![image](https://user-images.githubusercontent.com/123335700/214119702-0a070b4b-477d-440f-b69e-88f2a12a176a.png)

**Comment by dieselrabbit** [comment]
The core problem is that all ScreenLogic messages that we've seen up until now were smaller than the buffer receiving the data so it was assumed that I'd get a complete SL message each time. Your 27 (total) circuits broke that. 🙂 But it's good. This'll only make it more robust.
I've implemented a buffer on the incoming data 6e42df771171fddbf66371ae6ac683c71e59df0c to wait for and re-assemble any split messages. This should fix the issue and I'm testing now.

**Comment by American1337** [comment]
Thanks for the quick fix !

I merged the change into my home assistant install. I tried a few of the circuits and they work :)  yay.

I'm unblocked for now, but I'm curious, when do you expect this patch to get merged into an official HA release? I will need to hold off installing official HA updates I'd expect.


![image](https://user-images.githubusercontent.com/123335700/214355642-5bd98547-7256-4fa2-becc-a8b5150a248c.png)

**Comment by American1337** [comment]
An interesting thing to note is only "U Pool Ligh" (upper pool light has the light icon) - the others show as switches. 

This isn't a problem to me - I can do my automations regardless of switch or light type, but looks like perhaps a glitch somewhere in parsing attributes.

The lights all show as lights in the app's light page.

![image](https://user-images.githubusercontent.com/123335700/214356383-d3ceb062-c963-4f28-8d94-f38bea17f501.png)
![image](https://user-images.githubusercontent.com/123335700/214356450-bc2d31c8-ab66-4fbc-a87c-6783972d539c.png)
![image](https://user-images.githubusercontent.com/123335700/214356613-14739d41-7d1f-494e-872d-3c9747fb6191.png)

**Comment by dieselrabbit** [comment]
> I'm unblocked for now, but I'm curious, when do you expect this patch to get merged into an official HA release? I will need to hold off installing official HA updates I'd expect.
> 

An issue will have to be created in the [home-assistant/core](https://github.com/home-assistant/core) repository so that I can bump the version of `screenlogicpy`. FYI, there are other updates coming for the screenlogic integration with the next monthly release on Feb 1st. Unsure, but it might be best to confirm this still happens with that release (it will) and create the issue as a bug for that release so hopefully it doesn't have to wait until the March release to be fixed.

**Comment by dieselrabbit** [comment]
> An interesting thing to note is only "U Pool Ligh" (upper pool light has the light icon) - the others show as switches.
> 

This is because HA is only turing circuits that are setup as a `Light` or `IntelliBright` into light entities. I hadn't seen a config using SAM, SAL, Photon Generator, or Color Wheel before so I wasn't able to confirm them. The fix for this will need to be made in Home Assistant though, so another issue will need to be created there.

For the lights in your configuration I'm reading:

1. Fiber Optic -> Photon Generator
2. Color Wheel -> Color Wheel
3. Color Wheel -> Color Wheel
4. Spa Light -> SAL Light
5. Pool Light -> ~~UNKNOWN~~ SAM Light
6. Pool Light -> SAM Light
7. Pool Light -> SAM Light
8. Pool Light -> SAM Light
9. Pool Light -> SAM Light
10. U Pool Ligh -> Light

Is that correct?
~~Additionally, I'm getting an unexpected value for the function of the first Pool Light. Could you confirm what circuit function it's set to?~~

Turns out markdown was messing with the data in the log you posted. Got that sorted and now it looks like all the lighting circuits show their expected circuit functions?

**Comment by American1337** [comment]
What you've decoded looks correct to me. 

What is the next step? Do I create an issue in https://github.com/home-assistant/core and reference both these issues?

**Comment by dieselrabbit** [comment]
Great!

Yes the next step would be to create two separate issues, one for each of these. 
As they are unrelated and HA is pretty particular about their issue tracking, they will likely ask you to separate them if submitted combined.

**Comment by American1337** [comment]
I filed both of them:

- https://github.com/home-assistant/core/issues/86593
- https://github.com/home-assistant/core/issues/86594

**Comment by American1337** [comment]
Thanks for quick fix on these issues! 🥇  I merged in the light fix (const.py), readded the integration and now the lights are now categorized correctly.

![image](https://user-images.githubusercontent.com/123335700/214601776-f0fde520-dc95-41be-a7f7-58b93c986eaf.png)

**Comment by dieselrabbit** [comment]
Glad everything is working. 

One note, you may have to manually delete the old `switch` entities for the lights.

Closing this as all is fixed and merged with HA.
